File: tools-manage/invocationlab-admin/src/main/java/org/javamaster/invocationlab/admin/filter/ResponseWrapper.java
package org.javamaster.invocationlab.admin.filter;


import javax.servlet.ServletOutputStream;
import javax.servlet.WriteListener;
import javax.servlet.http.HttpServletResponse;
import javax.servlet.http.HttpServletResponseWrapper;
import java.io.ByteArrayOutputStream;
import java.io.PrintWriter;
import java.nio.charset.StandardCharsets;

/**
 * @author yudong
 */
public class ResponseWrapper extends HttpServletResponseWrapper {

    private final ServletByteArrayOutputStream outputStream = new ServletByteArrayOutputStream();

    public ResponseWrapper(HttpServletResponse response) {
        super(response);
    }

    @Override
    public PrintWriter getWriter() {
        return new PrintWriter(outputStream);
    }

    @Override
    public ServletOutputStream getOutputStream() {
        return outputStream;
    }

    @Override
    public String toString() {
        return new String(outputStream.toByteArray(), StandardCharsets.UTF_8);
    }

}

class ServletByteArrayOutputStream extends ServletOutputStream {

    private final ByteArrayOutputStream buf = new ByteArrayOutputStream();

    @Override
    public boolean isReady() {
        return false;
    }

    @Override
    public void setWriteListener(WriteListener listener) {

    }

    @Override
    public void write(int b) {
        buf.write(b);
    }

    public byte[] toByteArray() {
        return buf.toByteArray();
    }
}

File: tools-manage/invocationlab-admin-21/src/main/kotlin/org/javamaster/invocationlab/admin/filter/ResponseWrapper.kt
package org.javamaster.invocationlab.admin.filter

import jakarta.servlet.ServletOutputStream
import jakarta.servlet.WriteListener
import jakarta.servlet.http.HttpServletResponse
import jakarta.servlet.http.HttpServletResponseWrapper
import java.io.ByteArrayOutputStream
import java.io.PrintWriter
import java.nio.charset.StandardCharsets


/**
 * @author yudong
 */
class ResponseWrapper(response: HttpServletResponse?) : HttpServletResponseWrapper(response) {
    private val outputStream = ServletByteArrayOutputStream()

    override fun getWriter(): PrintWriter {
        return PrintWriter(outputStream)
    }

    override fun getOutputStream(): ServletOutputStream {
        return outputStream
    }

    override fun toString(): String {
        return String(outputStream.toByteArray(), StandardCharsets.UTF_8)
    }
}

internal class ServletByteArrayOutputStream : ServletOutputStream() {
    private val buf = ByteArrayOutputStream()

    override fun isReady(): Boolean {
        return false
    }

    override fun setWriteListener(listener: WriteListener) {
    }

    override fun write(b: Int) {
        buf.write(b)
    }

    fun toByteArray(): ByteArray {
        return buf.toByteArray()
    }
}
